src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,464 @@
+-------------------------------------------------------------------------------}
+
+-- | A mutable version of 'CompactIndex'.
 --
--- This module includes in-memory parts and I\/O parts:
+-- Once frozen using 'freeze' or 'unsafeFreeze', a 'CompactIndex' can not be
+-- converted into a 'MCompactIndex' anymore.
 --
--- * Incremental construction
+data MCompactIndex s = MCompactIndex {
+    -- * Core index structure
+    mciRangeFinder          :: !(VU.MVector s Word32)
+  , mciRangeFinderPrecision :: !Int
+  , mciPrimary              :: !(VU.MVector s Word32)
+  , mciClashes              :: !(VU.MVector s Bit)
+  , mciTieBreaker           :: !(STRef s (IntMap SerialisedKey))
+    -- * Aux information required for incremental construction
+
+    -- | The number of the current disk page we are constructing the index for.
+  , mciCurrentPageNumber    :: !(STRef s Int)
+    -- | The range-finder bits of the page-minimum key that we saw last.
+    --
+    --  This should be 'SNothing' if we haven't seen any keys/pages yet.
+  , mciLastMinRfbits        :: !(STRef s (SMaybe Word16))
+    -- | The primary bits of the page-maximum key that we saw last.
+    --
+    -- This should be 'SNothing' if we haven't seen any keys/pages yet.
+  , mciLastMaxPrimbits      :: !(STRef s (SMaybe Word32))
+  }
+
+-- | One-shot construction.
 --
--- * Index lookup (index search)
+fromList :: Int -> [(SerialisedKey, SerialisedKey)] -> CompactIndex
+fromList rfprec ks = runST $ unsafeFreeze =<< fromList' rfprec ks
+
+fromList' :: Int -> [(SerialisedKey, SerialisedKey)] -> ST s (MCompactIndex s)
+fromList' rfprec ks = do
+    mci <- new rfprec n
+    mapM_ (`append` mci) ks
+    pure mci
+  where
+    n = length ks
+
 --
--- * (de-)serialisation
+-- Incremental construction
 --
--- The above list is a sketch. Functionality may move around, and the list is
--- not exhaustive.
+
+{- $incremental
+
+  Constructing a compact index incrementally can go wrong, unless the following
+  preconditions are met and invariants upheld.
+
+  * /Non-empty/: a compact index can not be empty. A mutable compact index can
+    be empty at first, but at least one page should be appended before it is
+    frozen.
+
+  * /Sorted/: pages must be appended in sorted order according to the keys they
+    contain.
+
+  * /Unique/: all keys should be unique within and across pages. A page can
+    contain only one key, in which case the minimum and maximum keys would
+    match.
+
+  * /Partitioned/: pages must be partitioned. That is, the range-finder bits for
+    all keys within a page must match.
+-}
+
+-- | @'new' rfprec n@ creates a new mutable index with a range-finder
+-- bit-precision of @rfprec, and with space reserved for @n@ (disk page)
+-- entries.
 --
-module Database.LSMTree.Internal.Run.Index.Compact () where
+-- Note: after initialisation, the (mutable) index is not resizable, and the
+-- range-finder bit-precision can not be changed.
+new :: Int -> Int -> ST s (MCompactIndex s)
```

Comment:
What's the purpose of having both the bulk and incremental versions? Are we expecting to use the bulk version for writing out the write buffer? Is it faster?

Or is it just there as a reference for testing? If the latter, it could live in the testsuite and doens't need to be written in an efficient way, but as simple as possible.